manager/integration/tests/test_upgrade.py

```diff
@@ -45,39 +62,111 @@ def test_upgrade(upgrade_image_tag, volume_name):  # NOQA
     new_ei_name = "longhornio/longhorn-engine:" + upgrade_image_tag
 
     client = get_longhorn_api_client()
-
+    core_api = get_core_api_client()
     host_id = get_self_host_id()
+    pod_data_path = "/data/test"
+
+    pod_volume_name = generate_volume_name()
+
+    # Create Volume attached to a node.
+    volume1 = create_and_check_volume(client,
+                                      volume_name,
+                                      size=SIZE)
+    volume1.attach(hostId=host_id)
+    volume1 = wait_for_volume_healthy(client, volume_name)
+    volume1_data = write_volume_random_data(volume1)
+
+    # Create Volume used by Pod
+    pod_name, pv_name, pvc_name, pod_md5sum = \
+        prepare_pod_with_data_in_mb(client, core_api,
+                                    pod_make, pod_volume_name,
+                                    data_path=pod_data_path,
+                                    add_liveness_prope=False)
+
+    # Create multiple volumes used by StatefulSet
+    statefulset_name = 'statefulset-upgrade-test'
+    update_statefulset_manifests(statefulset,
+                                 storage_class,
+                                 statefulset_name)
+    create_storage_class(storage_class)
+    create_and_wait_statefulset(statefulset)
+    statefulset_pod_info = get_statefulset_pod_info(core_api, statefulset)
+
+    # upgrade Longhorn
+    assert longhorn_upgrade(upgrade_image_tag)
 
-    volume = create_and_check_volume(client,
-                                     volume_name,
-                                     size=SIZE)
+    client = get_longhorn_api_client()
 
-    volume.attach(hostId=host_id)
+    # wait for 1 minute before checking pod restarts
+    time.sleep(60)
 
-    volume = wait_for_volume_healthy(client, volume_name)
+    pod = core_api.read_namespaced_pod(name=pod_name,
+                                       namespace='default')
+    assert pod.status.container_statuses[0].restart_count == 0
 
-    data = write_volume_random_data(volume)
+    for sspod_info in statefulset_pod_info:
+        sspod = core_api.read_namespaced_pod(name=sspod_info['pod_name'],
+                                             namespace='default')
+        assert \
+            sspod.status.container_statuses[0].restart_count == 0
 
-    assert longhorn_upgrade(upgrade_image_tag)
+    statefulset['spec']['replicas'] = replicas = 0
+    apps_api = get_apps_api_client()
 
-    client = get_longhorn_api_client()
+    apps_api.patch_namespaced_stateful_set(
+        name=statefulset_name,
+        namespace='default',
+        body={
+            'spec': {
+                'replicas': replicas
+            }
+        })
 
-    volume = client.by_id_volume(volume_name)
+    delete_and_wait_pod(core_api, pod_name)
```

Comment:
Have we checked the pod's data?

Reply:
Updated it to check data before/after engine upgrade